os/arch/arm/src/amebasmart/amebasmart_spi.c

```diff
@@ -1157,54 +1157,93 @@ static void amebasmart_spi_exchange(FAR struct spi_dev_s *dev,
  ************************************************************************************/
 
 	int mode_16bit = 1;
+	uint32_t len = 0;
 	spi_hookmastercallback(priv);
 	if (amebasmart_spi_9to16bitmode(priv)) {
 		/* 16-bit mode */
 		mode_16bit++;
 	}
-	DEBUGASSERT((nwords * mode_16bit) <= SPI_DMA_MAX_BUFFER_SIZE);
-	if (txbuffer && rxbuffer) {
-		uint8_t *rxbuff_aligned = NULL;
+	len = nwords * mode_16bit;
+	uint8_t *rxbuff_aligned = NULL;
+	uint8_t *txbuff_aligned =  NULL;
+	if (len >= SPI_DMA_MAX_BUFFER_SIZE) {
 		rxbuff_aligned = (uint8_t *)rtw_zmalloc(SPI_DMA_MAX_BUFFER_SIZE);
 		if(rxbuff_aligned == NULL) {
 			return;
 		}
-		uint8_t *txbuff_aligned =  NULL;
-		txbuff_aligned = (uint8_t *)rtw_zmalloc(SPI_DMA_MAX_BUFFER_SIZE);
-		if(txbuff_aligned == NULL) {
-			rtw_mfree(rxbuff_aligned, 0);
-			return;
+		if (txbuffer) {
+			txbuff_aligned = (uint8_t *)rtw_zmalloc(SPI_DMA_MAX_BUFFER_SIZE);
+			if(txbuff_aligned == NULL) {
+				rtw_mfree(rxbuff_aligned, 0);
+				return;
+			}
 		}
-		memcpy(txbuff_aligned, txbuffer, nwords * mode_16bit);
-		spi_master_write_read_stream_dma(&priv->spi_object, (char *)txbuff_aligned, (char *)rxbuff_aligned, (uint32_t)nwords * mode_16bit);
-		spi_dmarxwait(priv);
-		spi_dmatxwait(priv);
-		memcpy(rxbuffer, rxbuff_aligned, nwords * mode_16bit);
-		rtw_mfree(rxbuff_aligned, 0);
-		rtw_mfree(txbuff_aligned, 0);
-	} else if (txbuffer) {
-		uint8_t *txbuff_aligned =  NULL;
-		txbuff_aligned = (uint8_t *)rtw_zmalloc(SPI_DMA_MAX_BUFFER_SIZE);
-		if(txbuff_aligned == NULL) {
-			return;
+		while (len >= SPI_DMA_MAX_BUFFER_SIZE) {
```

Comment:
```
state = 0;
if (txbuffer) {
   alloc txbuff_alinged with SPI_DMA_MAX_BUFFER_SIZE;
   state += 1;
}
if (rxbuffer) {
  alloc rxbuff_aligned with SPI_DMA_MAX_BUFFER_SIZE;
  state += 2;
}

while (remain_data_len > 0) {
   if (remain_data_len >= SPI_DMA_MAX_BUFFER_SIZE) {
      send_data_len = SPI_DMA_MAX_BUFFER_SIZE;
   } else {
      send_data_len = remain_data_len;
   }

  switch (state) {
  case 3: // tx+rx
    do r/tx;
    break;
  case 2: // rx only
     do rx only
     break;
  case 1: // tx only
      do tx only
      break;
   case 0: // do nothing
      break;
   }

   remain_data_len -= send_data_len;
}

if (txbuff_aligned)
  free

if (rxbuff_aligned)
  free
```

It looks it has duplicated code, and even rxbuffer is null, it allocates rxbuff_aligned always.
How about above?

Reply:
rx buffer is required as we asked @zhongnuo-tang to remove the delay (0.5ms) after spi tx transfer. Delay was slower when we compared it using read_write (both tx and rx) for tx.

Reply:
Hi @sunghan-chang,
I have removed duplicate code as suggested, and rxbuffer is required like what Abhishek had explained,